### PR TITLE
add v2rayN

### PR DESCRIPTION
**Pull request by linsui**
GUI client based on v2ray core

### CODE REVIEW

File: bucket/v2rayn.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        "    Write-Host \"Creating hardlink for $file\"",
+        "        New-Item -ItemType HardLink -Path $dir -Name $file -Target \"$(appdir v2ray)\\current\\$file\"",
+        "    }",
+        "}} | Out-Null"
+    ],
+    "checkver": "github",
+    "autoupdate": {
+        "url": "https://github.com/2dust/v2rayN/releases/download/$version/v2rayN.zip"
+    }
+}
```
Comment (Ash258) [inline bucket/v2rayn.json]: Missing new line

File: bucket/v2rayn.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    "depends": "v2ray",
+    "shortcuts": [
+        [
+            "v2rayN.exe",
+            "v2rayN"
+        ]
+    ],
+    "pre_install": [
+        "if (!(Test-Path \"$persist_dir\\user-wininet.json\")) {",
+        "    New-Item -Force -Path \"$persist_dir\" -Name 'user-wininet.json' -Value '{}' | Out-Null",
```
Comment (Ash258) [inline bucket/v2rayn.json]:
```suggestion
        "    New-Item -Force -Path \"$dir\" -Name 'user-wininet.json' -Value '{}' | Out-Null",
```

File: bucket/v2rayn.json [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        "}",
+        "if (!(Test-Path \"$persist_dir\\config.json\")) {",
+        "    Copy-Item \"$(appdir v2ray)\\current\\config.json\" $persist_dir | Out-Null",
```
Comment (Ash258) [inline bucket/v2rayn.json]:
```suggestion
        "    Copy-Item \"$(appdir v2ray $global)\\current\\config.json\" \"$dir\" | Out-Null",
```

File: bucket/v2rayn.json [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        "}",
+        "'*.exe', '*.dat', 'vpoint*.json' | % { % ($file in dir \"$(appdir v2ray)\\current\") {",
```
Comment (Ash258) [inline bucket/v2rayn.json]: This syntax is not readable at all. 

What is $_ in this context?
Convert it to something like:

```powershell
foreach $form in @(..formats) {
    current_dir \"$dir\" | ForEach-Object {
        if $_ -like $form { .....
    }
}
```

File: bucket/v2rayn.json [hunk 5]
```diff
@@ -20,20 +20,22 @@
     ],
     "pre_install": [
         "if (!(Test-Path \"$persist_dir\\user-wininet.json\")) {",
-        "    New-Item -Force -Path \"$dir\" -Name 'user-wininet.json' -Value '{}' | Out-Null",
+        "    New-Item -Force -Path $dir -Name 'user-wininet.json' -Value '{}' | Out-Null",
```
Comment (Ash258) [inline bucket/v2rayn.json]: Again??
```suggestion
        "    New-Item -Force -Path \"$dir\" -Name 'user-wininet.json' -Value '{}' | Out-Null",
```

File: bucket/v2rayn.json [hunk 6]
```diff
@@ -20,20 +20,22 @@
...
         "}",
         "if (!(Test-Path \"$persist_dir\\config.json\")) {",
-        "    Copy-Item \"$(appdir v2ray $global)\\current\\config.json\" \"$dir\" | Out-Null",
+        "    Copy-Item \"$(appdir v2ray $global)\\current\\config.json\" $dir | Out-Null",
```
Comment (Ash258) [inline bucket/v2rayn.json]:
```suggestion
        "    Copy-Item \"$(appdir v2ray $global)\\current\\config.json\" \"$dir\" | Out-Null",
```

File: bucket/v2rayn.json [hunk 7]
```diff
@@ -20,20 +20,22 @@
...
         "}",
-        "'*.exe', '*.dat', 'vpoint*.json' | % { % ($file in dir \"$(appdir v2ray)\\current\") {",
-        "    if ($file -Like $_){",
-        "    Write-Host \"Creating hardlink for $file\"",
-        "        New-Item -ItemType HardLink -Path $dir -Name $file -Target \"$(appdir v2ray)\\current\\$file\"",
+        "foreach ($form in @('*.exe', '*.dat', 'vpoint*.json')) {",
+        "    foreach ($file in dir \"$(appdir v2ray)\\current\") {",
```
Comment (Ash258) [inline bucket/v2rayn.json]:
```suggestion
        "    foreach ($file in Get-ChildItem \"$(appdir v2ray $global)\\current\" -File) {",
```

File: bucket/v2rayn.json [hunk 8]
```diff
@@ -20,20 +20,22 @@
...
+        "        if ($file -Like $form) {",
+        "            Write-Host \"Creating hardlink for $file\"",
+        "            New-Item -ItemType HardLink -Path $dir -Name $file -Target \"$(appdir v2ray)\\current\\$file\" | Out-Null",
```
Comment (Ash258) [inline bucket/v2rayn.json]:
```suggestion
        "            New-Item -ItemType HardLink -Path \"$dir\" -Name $file -Target \"$(appdir v2ray $global)\\current\\$file\" | Out-Null",
```

**Comment by linsui** [comment]
@Ash258  Could you tell me when should I use `$global` in `$(appdir app $global)` ? :confused:

**Comment by Ash258** [comment]
Always. Global installation would fail otherwise

https://github.com/lukesampson/scoop-extras/pull/1869#issue-261685998

**Comment by linsui** [comment]
Thanks. Is this all right now?

**Comment by Ash258** [comment]
I guess. I am not sure how powershell will handle that `$file` variable inside script. Maybe it should be qutoed to preserve spaces. But i don't know exact flow of "replacing" and evaluating variables. Need to test how `Invoke-Expression` will handle it.

**Comment by linsui** [comment]
Maybe I should change it to `$name` ? And can't alias be used? Is foreach OK? I hope there is a more detailed wiki :joy:

**Comment by Ash258** [comment]
Usage of aliases is not good idea overall while writing scripts. When you are writing commands in terminal, then yeah, go for aliases. But when you write code / manifests it need to be straightforward and easy to read / edit. 

https://github.com/PowerShell/PSScriptAnalyzer/blob/development/RuleDocumentation/AvoidUsingCmdletAliases.md

image

**Comment by linsui** [comment]
Is this all right?

**Comment by Ash258** [comment]
`foreach` is normal keyword in powershell which is totally OK.  Also `foreach` and `foreach-object` behaves differently (based on it's usage). https://stackoverflow.com/a/52230828/6120775

**Comment by linsui** [comment]
It's really confusing.

**Comment by linsui** [comment]
Is this all right now?

**Pull request merged**